Admin - Dev Mode: grey out unavailable modules. Make items in At A Glance unclickable and faded

eliorivero opened the pull request:
Fixes #4006.

#### Changes proposed in this Pull Request:
- if a module isn't available in dev mode, its toggle is hidden and the legend _Unavailable in Dev Mode_ is displayed. They receive the CSS class `devmode-disabled`.
<img width="732" alt="module" src="https://cloud.githubusercontent.com/assets/1041600/15719244/7280b6aa-2805-11e6-827e-bda92033675c.png">

- if an element in At A Glance requires connection to work and site is in dev mode, it will display the legend _Unavailable in Dev Mode_
<img width="753" alt="dashboard" src="https://cloud.githubusercontent.com/assets/1041600/15719245/728d1ea4-2805-11e6-82ba-edd4b09e3d37.png">

- Makes some strings [available for translation](https://github.com/Automattic/jetpack/pull/4026/files#diff-7d41d2825e41836777e5ff2b874d9099R34) that were missing in `protect.jsx`

#### New functions
- `isDevMode`: receives the global state tree, returns a boolean value stating whether the site is in dev mode or not
- `requiresConnection`: receives a slug-like string, for example, a module slug, and checks if it's in a list of elements that require connection.
- `isUnavailableInDevMode`: receives the global state tree and a slug-like string, and checks if the site is in dev mode and if the element requires connection to be available.
- `getModulesThatRequireConnection`: receives the global state tree and returns all modules that require a connection.

#### Component Update
- DashItem now accepts a `disabled` property that makes it unclickable and faded.

dereksmart commented on _inc/client/state/connection/reducer.js:
```diff
@@ -180,3 +180,57 @@ export function isFetchingUserData( state ) {
+ */
+export function requiresConnection( slug ) {
+	return [
+		// Modules
```
Couldn't we get these programmatically from getModules()?  Each module has a `requires_connection` value.

eliorivero replied:
Nice, I'll build something around it.

eliorivero replied:
That said, there are a couple of cards that aren't actual modules, like backups and scan so I'll have to add a fallback for those.

dereksmart replied:
this list isn't going to change anytime in the near future, and since it's just dev mode honestly I'm ok with a hardcoded version for this time around.

eliorivero replied:
@dereksmart I've changed this now for a function that will get the values from the modules.

dereksmart commented on _inc/client/more/index.jsx:
```diff
@@ -18,8 +18,14 @@ import {
 
-export const Page = ( { toggleModule, isModuleActivated, isTogglingModule, getModule } ) => {
+export const Page = ( props ) => {
+	let { toggleModule,
```
Personally I think it looks better like 

```
let {
    toggleModule
    ...
} = props;
```

dereksmart commented on _inc/client/security/index.jsx:
```diff
@@ -18,8 +18,14 @@ import {
 
-export const Page = ( { toggleModule, isModuleActivated, isTogglingModule, getModule } ) => {
+export const Page = ( props ) => {
+	let { toggleModule,
```
same nitpick about code style

dereksmart commented on _inc/client/site-health/index.jsx:
```diff
@@ -17,8 +17,14 @@ import {
 
-export const Page = ( { toggleModule, isModuleActivated, isTogglingModule, getModule } ) => {
+export const Page = ( props ) => {
+	let { toggleModule,
```
same nitpick about code style

dereksmart commented:
Aside from a couple of code style nitpicks, this LGTM code-wise.  We'll need a design review eventually.

eliorivero commented:
@dereksmart it's all done now. Added a couple of screenshots showing how it looks.

@jeffgolenski I used the stylesheet that we were talking about earlier to add a couple of styles for the module foldable card when it's unavailable in dev mode. I added it in a separate commit https://github.com/Automattic/jetpack/pull/4026/commits/1a51a0abce710e438d6125678ef4a290582d7793 so I can quickly remove it if you don't want it there.
